lib/dpl/provider/pypi.rb

```diff
@@ -32,23 +32,13 @@ def skip_upload_docs?
           (options.has_key?(:skip_upload_docs) && options[:skip_upload_docs])
       end
 
-      def self.install_setuptools
-        shell 'wget https://bootstrap.pypa.io/ez_setup.py -O - | sudo python'
-        shell 'rm -f setuptools-*.zip'
-      end
-
-      def self.install_twine
-        shell("pip install twine", retry: true) if `which twine`.chop.empty?
-      end
-
-      def initialize(*args)
-        super(*args)
-        self.class.pip 'wheel' if pypi_distributions.to_s.include? 'bdist_wheel'
+      def install_deploy_dependencies
+        unless context.shell 'wget https://bootstrap.pypa.io/get-pip.py -O - | sudo python'
+          error "Couldn't install pip and setuptools."
+        end
+        context.shell("pip install twine", retry: true) if `which twine`.chop.empty?
```

Comment:
Why not upgrade pip and twine to latest versions at this point?

Reply:
`get-pip.py` will already [install the latest versions of setuptools, wheel and pip](https://github.com/pypa/get-pip/blob/430ba37776ae2ad89f794c7a43b90dc23bac334c/template.py#L170-L172) itself and running `pip install twine` effectively means installing the most recent version from PyPI.

Reply:
(comment edited to correct line numbers in source link)

Reply:
Good catch, so I'd suggest to do:
```
get-pip.py --no-setuptools --no-wheel
pip install --upgrade setuptools twine wheel
```

Reply:
I could have misread, but I think this would work too:
`get-pip.py setuptools wheel twine`

Reply:
Though your way is less reliant on get-pip.py quirks so perhaps best.

Reply:
You're right, I was confused by how the `implicit_*` flag where handled and that it would take sys.argv straight to do the call with `--upgrade`. The code around this is slightly weird to make sure that there isn't a double entry of setuptools or wheel in the args.